cdap-event-reader-spi/src/main/java/io/cdap/cdap/spi/events/PubSubEventReader.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.spi.events;
+
+import java.util.List;
+
+/**
+ * {@link PubSubEventReader} Interface for listening for events from Pub/Sub.
+ */
+public interface PubSubEventReader<T extends Event> extends AutoCloseable {
+  /**
+   * Method to initialize PubSubEventReader.
+   *
+   * @param context Configuration properties of reader
+   */
+  void initialize(PubSubEventReaderContext context);
+
+  /**
+   * Pull messages from PubSub if available.
+   *
+   * @param maxMessages maximum messages to pull
+   * @return List of Messages
+   */
+  List<T> pull(int maxMessages);
+
+  /**
+   * Action to perform on successful processing of event.
+   *
+   * @param ackId Ack Id of Message to acknowledge
+   * @throws Exception Invalid ACK ID / Expired ACK ID
+   */
+  void success(String ackId) throws Exception;
```

Comment:
Also, ackId is very specific to pubsub, making this interface not generic. It is better to have the `pull` method return an object that the caller can use to acknowledge. E.g.

```java
EventResult<T> pull(int max);

...
interface EventResult<T> extends AutoCloseable {
  // Implementation details. Consume all messages. If exception is raised from the Consumer.accept() method for a particular message, it will not be ack upon closing of the event result.
  void consumeMessages(Consumer<T> consumer)
}
```